packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -448,7 +448,21 @@ class TypeMatcher<T> {
 /// updating the underlying element (i.e., by calling [Element.update] with the
 /// new widget). Otherwise, the old element is removed from the tree, the new
 /// widget is inflated into an element, and the new element is inserted into the
-/// tree.
+/// tree. Every time you are building a list of widgets from a list of data
+/// and the list might change, you should set a key to the children, so that
+/// Flutter keep the subtrees tied to the data. Otherwise, Flutter will not know
+/// what is going on and if any widget is stateful the data could spill from one
+/// widget state to the other, because Flutter has no way to distinguish
+/// "an item changed its data" from "one of the items was removed/added and
+/// pushed all the other items around" without keys).
```

Comment:
Not sure if we should keep "Flutter will not know what is going on". Instead we can use something like below to precisely explain using specific terms:
"Otherwise if a widget is stateful without key property set to its children, the data could get lost in widget sub-tree".
Also the last two lines could be re-phrased.